File: papers/vpt/originalFirstNando.ml
open GT
open OCanren
open OCanren.Std
open Helper

let topLevel x0 x1 =
  let rec evalo st fm u = (fresh (x y v w z) ((((fm === (conj x y)) &&& ((ando v w u) &&& ((evalo st x v) &&& (evalo st y w)))) ||| (((fm === (disj x y)) &&& ((oro v w u) &&& ((evalo st x v) &&& (evalo st y w)))) ||| (((fm === (neg x)) &&& ((noto v u) &&& (evalo st x v))) ||| ((fm === (var z)) &&& (elemo z st u)))))))
  and ando a b c = (fresh (ab) (((nando a b ab) &&& (nando ab ab c))))
  and nando a b c = (((a === !!false) &&& ((b === !!false) &&& (c === !!true))) ||| (((a === !!false) &&& ((b === !!true) &&& (c === !!true))) ||| (((a === !!true) &&& ((b === !!false) &&& (c === !!true))) ||| ((a === !!true) &&& ((b === !!true) &&& (c === !!false))))))
  and oro a b c = (fresh (aa bb) (((nando a a aa) &&& ((nando b b bb) &&& (nando aa bb c)))))
  and noto a na = (nando a a na)
  and elemo n s v = (fresh (h t n') ((((n === Nat.zero) &&& ((s === (h % t)) &&& (v === h))) ||| ((n === (Nat.succ n')) &&& ((s === (h % t)) &&& (elemo n' t v))))))
  in         (((evalo x0 x1 !!true)))

File: papers/vpt/originalFirstNando.pl
evalo(St, conj(X, Y), U) :- ando(V, W, U), evalo(St, X, V), evalo(St, Y, W).
evalo(St, disj(X, Y), U) :- oro(V, W, U), evalo(St, X, V), evalo(St, Y, W).
evalo(St, neg(X), U) :- noto(V, U), evalo(St, X, V).
evalo(St, var(Z), U) :- elemo(Z, St, U).
ando(A, B, C) :- nando(A, B, Ab), nando(Ab, Ab, C).
nando(false, false, true).
nando(false, true, true).
nando(true, false, true).
nando(true, true, false).
oro(A, B, C) :- nando(A, A, Aa), nando(B, B, Bb), nando(Aa, Bb, C).
nando(false, false, true).
nando(false, true, true).
nando(true, false, true).
nando(true, true, false).
noto(A, Na) :- nando(A, A, Na).
nando(false, false, true).
nando(false, true, true).
nando(true, false, true).
nando(true, true, false).
elemo(o, cons(H, T), H).
elemo(s(N_0), cons(H, T), V) :- elemo(N_0, T, V).